crates/runtime/src/memfd.rs

```diff
@@ -0,0 +1,566 @@
+        // termination.  The `clear_and_remain_ready()` path also
+        // mprotects memory above the initial heap size back to
+        // PROT_NONE, so we don't need to do that here.
+        if (self.image.is_none()
+            && maybe_image.is_none()
+            && self.initial_size == initial_size_bytes)
+            || (self.image.is_some()
+                && maybe_image.is_some()
+                && self.image.as_ref().unwrap().fd.as_file().as_raw_fd()
+                    == maybe_image.as_ref().unwrap().fd.as_file().as_raw_fd())
+        {
+            self.dirty = true;
+            self.cur_size = initial_size_bytes;
+            return Ok(());
+        }
+
+        // Otherwise, we need to redo (i) the anonymous-mmap backing
+        // for the whole slot, (ii) the initial-heap-image mapping if
+        // present, and (iii) the mprotect(PROT_NONE) above the
+        // initial heap size.
+
+        // Security/audit note: we map all of these MAP_PRIVATE, so
+        // all instance data is local to the mapping, not propagated
+        // to the backing fd. We throw away this CoW overlay with
+        // madvise() below, from base up to static_size (which is the
+        // whole slot) when terminating the instance.
+
+        // Anonymous mapping behind the initial heap size: this gives
+        // zeroes for any "holes" in the initial heap image. Anonymous
+        // mmap memory is faster to fault in than a CoW of a file,
+        // even a file with zero holes, because the kernel's CoW path
+        // unconditionally copies *something* (even if just a page of
+        // zeroes). Anonymous zero pages are fast: the kernel
+        // pre-zeroes them, and even if it runs out of those, a memset
+        // is half as expensive as a memcpy (only writes, no reads).
+        //
+        // We map these inaccessible at first then mprotect() the
+        // whole of the initial heap size to R+W below.
+        //
+        // Special case: we can skip if the last instantiation had no
+        // image. This means that the whole slot is filled with an
+        // anonymous mmap backing (and it will have already been
+        // cleared by the madvise). We may however need to
+        // mprotect(NONE) the space above `initial_size_bytes` if the
+        // last use of this slot left it larger. This also lets us
+        // skip an mmap the first time a MemFdSlot is used, because we
+        // require the caller to give us a fixed address in an
+        // already-mmaped-with-anon-memory region. This is important
+        // for the on-demand allocator.
+        if self.image.is_some() {
+            self.map_anon_memory(rustix::io::ProtFlags::empty())
+                .map_err(|e| InstantiationError::Resource(e.into()))?;
+        } else if initial_size_bytes < self.initial_size {
+            self.set_protection(
+                initial_size_bytes,
+                self.initial_size,
```

Comment:
I think this should be `self.initial_size - initial_size_bytes`, right?

In that this is just making the previously-larger heap inaccessible after the base?

Reply:
I think it's correct, but it's pretty subtle so I added some more comments with a figure showing the before-and-after mmap state. Basically we have 0..self.initial_size accessible, and we want only 0..initial_size_bytes (the new, smaller size) accessible, so we make inaccessible the range initial_size_bytes..self.initial_size.

Reply:
But this parameter is the length parameter, not an end, so isn't the length of what we're changing to PROT_NONE `self.initial_size - initial_size_bytes` since the new heap is smaller?